### PR TITLE
Introduce ds session interaction with shared subs

### PR DESCRIPTION
**Pull request by savonarola**
**DO NOT MERGE**

Fixes [EMQX-12305](https://emqx.atlassian.net/browse/EMQX-12305)

[EMQX-12305]: https://emqx.atlassian.net/browse/EMQX-12305?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx/src/emqx_persistent_session_shared_sub/emqx_persistent_session_ds_shared_subs.erl [hunk 1]
```diff
@@ -0,0 +1,398 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx/src/emqx_persistent_session_shared_sub/emqx_persistent_session_ds_shared_subs.erl]: ✅

File: apps/emqx/src/emqx_persistent_session_shared_sub/emqx_persistent_session_ds_shared_subs.erl
Comment (savonarola) [inline apps/emqx/src/emqx_persistent_session_shared_sub/emqx_persistent_session_ds_shared_subs.erl:1]: When we are ok with the API, I plan to
* move all the actual code out of the module
* make static dispatching of all the methods
  * for CE app — to a module that forbids shared subs and always returns empty results
  * for the Platform — to a real module inside a separate BSL app

File: apps/emqx/include/emqx_session.hrl [hunk 2]
```diff
@@ -20,4 +20,6 @@
 -define(IS_SESSION_IMPL_MEM(S), (is_tuple(S) andalso element(1, S) =:= session)).
 -define(IS_SESSION_IMPL_DS(S), (is_map_key(id, S))).
 
+-define(session_message(MSG), {'$sess', MSG}).
```
Comment (ieQu1) [inline apps/emqx/include/emqx_session.hrl:23]: Make it a record?

File: apps/emqx/src/emqx_persistent_session_ds.hrl [hunk 3]
```diff
@@ -83,4 +84,8 @@
 -define(protocol, protocol).
 -define(offline_info, offline_info).
 
+-define(owner_session, owner_session).
+-define(owner_shared_sub, owner_shared_sub).
+
+-define(shared_sub_message(MSG), {'$share', MSG}).
```
Comment (ieQu1) [inline apps/emqx/src/emqx_persistent_session_ds.hrl:90]: Make it a record as well? Or a map.

File: apps/emqx/src/emqx_persistent_session_ds_subs.erl [hunk 4]
```diff
@@ -244,7 +246,14 @@ to_map(S) ->
 ) ->
     Acc.
 fold(Fun, Acc, S) ->
-    emqx_persistent_session_ds_state:fold_subscriptions(Fun, Acc, S).
+    emqx_persistent_session_ds_state:fold_subscriptions(
```
Comment (ieQu1) [inline apps/emqx/src/emqx_persistent_session_ds_subs.erl:249]: Hiding data from a generic function like `fold` will lead to confusion. Filtering shared subs is better done by business logic.

Comment (savonarola) [reply apps/emqx/src/emqx_persistent_session_ds_subs.erl:249]: Do you mean it is the callers who should decide which subscriptions to filter?

**Comment by savonarola** [comment]
Superseded by https://github.com/emqx/emqx/pull/13163

**Pull request closed without merging**